crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,307 @@
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        snapshot: &PartitionSnapshotMetadata,
+        local_snapshot_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        let partition_id = snapshot.partition_id;
+        let lsn = snapshot.min_applied_lsn;
+
+        debug!(
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        let relative_snapshot_path = format!("lsn_{lsn}", lsn = snapshot.min_applied_lsn);
+        let snapshot_prefix = format!(
+            "{prefix}{partition_id}/{relative_snapshot_path}",
+            prefix = self.prefix,
+        );
+
+        debug!(
+            "Uploading snapshot from {:?} to {}",
+            local_snapshot_path.as_path(),
+            snapshot_prefix
+        );
+
+        for file in &snapshot.files {
+            let filename = file.name.trim_start_matches("/");
+            let key = object_store::path::Path::from(format!(
+                "{}/{}",
+                snapshot_prefix.as_str(),
+                filename
+            ));
+            let put_result = put_snapshot_object(
+                local_snapshot_path.join(filename).as_path(),
+                &key,
+                &self.object_store,
+            )
+            .await?;
+            debug!(
+                etag = put_result.e_tag.unwrap_or_default(),
+                ?key,
+                "Put snapshot data file completed",
+            );
+        }
+
+        let metadata_key =
+            object_store::path::Path::from(format!("{}/metadata.json", snapshot_prefix.as_str()));
+        let metadata_json_payload = PutPayload::from(
+            serde_json::to_string_pretty(snapshot).expect("Can always serialize JSON"),
+        );
+        let put_result = self
+            .object_store
+            .put(&metadata_key, metadata_json_payload)
+            .await?;
+        debug!(
+            etag = put_result.e_tag.unwrap_or_default(),
+            key = ?metadata_key,
+            "Successfully published snapshot metadata",
+        );
+
+        let latest = LatestSnapshot {
+            version: snapshot.version,
+            lsn: snapshot.min_applied_lsn,
+            partition_id,
+            node_name: snapshot.node_name.clone(),
+            created_at: snapshot.created_at.clone(),
+            snapshot_id: snapshot.snapshot_id,
+            min_applied_lsn: snapshot.min_applied_lsn,
+            path: relative_snapshot_path,
+        };
+        let latest_path = object_store::path::Path::from(format!(
+            "{prefix}{partition_id}/latest.json",
+            prefix = self.prefix,
+            partition_id = partition_id,
+        ));
+        let latest_json_payload = PutPayload::from(serde_json::to_string_pretty(&latest)?);
+        let put_result = self
+            .object_store
+            .put(&latest_path, latest_json_payload)
+            .await?;
```

Comment:
Since I wrote that comment, we have been blessed with proper S3 conditional put, so I rewrote the update logic to perform a CAS 🎉 I'm not doing this preemptively since this path should be uncontended, but the check is there as a defensive measure against going backwards and overwriting something we didn't mean to.

Reply:
Sounds good :-)